timemachine/cpp/src/nonbonded_interaction_group.cu

```diff
@@ -0,0 +1,480 @@
+
+    gpuErrchk(cudaFreeHost(p_ixn_count_));
+
+    gpuErrchk(cudaFree(d_nblist_x_));
+    gpuErrchk(cudaFree(d_nblist_box_));
+    gpuErrchk(cudaFree(d_rebuild_nblist_));
+    gpuErrchk(cudaFreeHost(p_rebuild_nblist_));
+};
+
+template <typename RealType, bool Interpolated>
+void NonbondedInteractionGroup<RealType, Interpolated>::set_nblist_padding(double val) {
+    nblist_padding_ = val;
+}
+
+template <typename RealType, bool Interpolated>
+void NonbondedInteractionGroup<RealType, Interpolated>::disable_hilbert_sort() {
+    disable_hilbert_ = true;
+}
+
+template <typename RealType, bool Interpolated>
+void NonbondedInteractionGroup<RealType, Interpolated>::hilbert_sort(
+    const int N,
+    const unsigned int *d_atom_idxs,
+    const double *d_coords,
+    const double *d_box,
+    unsigned int *d_perm,
+    cudaStream_t stream) {
+
+    const int tpb = 32;
+    const int B = ceil_divide(N, tpb);
+
+    k_coords_to_kv_gather<<<B, tpb, 0, stream>>>(
+        N, d_atom_idxs, d_coords, d_box, d_bin_to_idx_, d_sort_keys_in_, d_sort_vals_in_);
+
+    gpuErrchk(cudaPeekAtLastError());
+
+    cub::DeviceRadixSort::SortPairs(
+        d_sort_storage_,
+        d_sort_storage_bytes_,
+        d_sort_keys_in_,
+        d_sort_keys_out_,
+        d_sort_vals_in_,
+        d_perm,
+        N,
+        0,                            // begin bit
+        sizeof(*d_sort_keys_in_) * 8, // end bit
+        stream                        // cudaStream
+    );
+
+    gpuErrchk(cudaPeekAtLastError());
+}
+
+template <typename RealType, bool Interpolated>
+void NonbondedInteractionGroup<RealType, Interpolated>::execute_device(
+    const int N,
+    const int P,
+    const double *d_x,
+    const double *d_p,   // 2 * N * 3
+    const double *d_box, // 3 * 3
+    const double lambda,
+    unsigned long long *d_du_dx,
+    unsigned long long *d_du_dp,
+    unsigned long long *d_du_dl,
+    unsigned long long *d_u,
+    cudaStream_t stream) {
+
+    // (ytz) the nonbonded algorithm proceeds as follows:
+
+    // (done in constructor), construct a hilbert curve mapping each of the 256x256x256 cells into an index.
+    // a. decide if we need to rebuild the neighborlist, if so:
+    //     - look up which cell each particle belongs to, and its linear index along the hilbert curve.
+    //     - use radix pair sort keyed on the hilbert index with values equal to the atomic index
+    //     - resulting sorted values is the permutation array.
+    //     - permute lambda plane/offsets, coords
+    // b. else:
+    //     - permute new coords
+    // c. permute parameters
+    // d. compute the nonbonded interactions using the neighborlist
+    // e. inverse permute the forces, du/dps into the original index.
+    // f. u and du/dl is buffered into a per-particle array, and then reduced.
+    // g. note that du/dl is not an exact per-particle du/dl - it is only used for reduction purposes.
+
+    if (N != N_) {
+        std::cout << N << " " << N_ << std::endl;
```

Comment:
Remove? Also another print statement on 293

Reply:
Replaced the print statements with more informative errors both here and in `nonbonded_all_pairs.cu` in 653fe4d